lib/auth/machineid/machineidv1/bot_service.go

```diff
@@ -0,0 +1,677 @@
+		UserName:    authz.ClientUsername(ctx),
+		BotUserName: BotResourceName(bot.Metadata.Name),
+		RoleName:    BotResourceName(bot.Metadata.Name),
+		BotName:     bot.Metadata.Name,
+		RoleCount:   int64(len(bot.Spec.Roles)),
+	})
+	if err := bs.emitter.EmitAuditEvent(ctx, &apievents.BotCreate{
+		Metadata: apievents.Metadata{
+			Type: events.BotCreateEvent,
+			Code: events.BotCreateCode,
+		},
+		UserMetadata: authz.ClientUserMetadata(ctx),
+		ResourceMetadata: apievents.ResourceMetadata{
+			Name: bot.Metadata.Name,
+		},
+	}); err != nil {
+		bs.logger.WithError(err).Warn("Failed to emit BotCreate audit event.")
+	}
+
+	return bot, nil
+}
+
+func UpsertBot(ctx context.Context, backend Backend, bot *pb.Bot) (*pb.Bot, error) {
+	if err := validateBot(bot); err != nil {
+		return nil, trace.Wrap(err, "validating bot")
+	}
+	user, role, err := botToUserAndRole(bot)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting to resources")
+	}
+
+	// Copy in generation from existing user if exists
+	existingUser, err := backend.GetUser(ctx, BotResourceName(bot.Metadata.Name), false)
+	if err == nil {
+		if existingGeneration, ok := existingUser.GetLabel(types.BotGenerationLabel); ok {
+			meta := user.GetMetadata()
+			meta.Labels[types.BotGenerationLabel] = existingGeneration
+			user.SetMetadata(meta)
+		} else {
+			return nil, trace.BadParameter("unable to determine existing generation for bot due to missing label")
+		}
+	}
+
+	user, err = backend.UpsertUser(ctx, user)
+	if err != nil {
+		return nil, trace.Wrap(err, "upserting bot user")
+	}
+	role, err = backend.UpsertRole(ctx, role)
+	if err != nil {
+		return nil, trace.Wrap(err, "upserting bot role")
+	}
+
+	bot, err = botFromUserAndRole(user, role)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting from resources")
+	}
+	return bot, nil
+}
+
+func (bs *BotService) UpsertBot(ctx context.Context, req *pb.UpsertBotRequest) (*pb.Bot, error) {
+	_, err := authz.AuthorizeWithVerbs(
+		ctx, bs.logger, bs.authorizer, false, types.KindBot, types.VerbCreate, types.VerbUpdate,
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	bot, err := UpsertBot(ctx, bs.backend, req.Bot)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	bs.reporter.AnonymizeAndSubmit(&usagereporter.BotCreateEvent{
+		UserName:    authz.ClientUsername(ctx),
+		BotUserName: BotResourceName(bot.Metadata.Name),
+		RoleName:    BotResourceName(bot.Metadata.Name),
+		BotName:     bot.Metadata.Name,
+		RoleCount:   int64(len(bot.Spec.Roles)),
+	})
+	if err := bs.emitter.EmitAuditEvent(ctx, &apievents.BotCreate{
+		Metadata: apievents.Metadata{
+			Type: events.BotCreateEvent,
+			Code: events.BotCreateCode,
+		},
+		UserMetadata: authz.ClientUserMetadata(ctx),
+		ResourceMetadata: apievents.ResourceMetadata{
+			Name: bot.Metadata.Name,
+		},
+	}); err != nil {
+		bs.logger.WithError(err).Warn("Failed to emit BotCreate audit event.")
+	}
+
+	return bot, nil
+}
+
+func (bs *BotService) UpdateBot(
```

Comment:
Throughout the codebase we have these `UpdateX` functions which semantically mean "if this resource already exists, overwrite it completely (a-la `Upsert`), otherwise error". AFAICT it's very unusual to need this semantic:
If you're calling `UpdateX` it almost always means that you've just called `GetX`, changed a field or two, and are now updating the object. This makes the check for "if this resource already exists" superfluous -- you already know it exists because you just got it with `GetX`.

Note that HTTP doesn't have any such semantic:
- `CREATE`: `CreateX`
- `GET`: `GetX`
- `DELETE`: `DeleteX`
- `PUT`: `UpsertX`
- `PATCH`: no equivalent except an odd `UpdateX` like `UpdateRemoteCluster`
- `POST`: according to [the RFC](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.3):
    
    > The POST method requests that the target resource process the
   representation enclosed in the request according to the resource's
   own specific semantics.
   
   Arguably our `UpdateX` is this miscellaneous type, however as I pointed out previously it's an odd semantic to implement.
   
Given this line of reasoning, I think we should aim to avoid adding these `UpdateX` by default. Usually an `Upsert` seems to do the job. 

One loose end I have on this topic is: how do `Upsert`s behave with the new optimistic locking behavior? If a resource has a `Revision`, will it be checked and rejected if it doesn't match what's in the database on an `Upsert`? If yes, then `Upsert` already implements the current `Update` semantic (and `Update`s can be replaced with `Upsert`s). If that's not how it works, and `Upserts` just ignore `Revision`, then the current `Update`s could be seen as means of distributed synchronization to prevent a case like

1. My code calls `GetX`
2. Some other part of the system calls `DeleteX` on that object
3. My code updates a field and calls `UpdateX` on that object

However I don't think this was an intentional design pattern.

cc @rosstimothy @espadolini

Reply:
Fwiw: my implementation here is following the recently proposed RFD 153 on the design of resource APIs. So it may be worth following up with some of these concerns there - https://github.com/gravitational/teleport/pull/34103 .

I believe the long term goal is to try and eliminate the Upsert RPCs and encourage the use of the more specific Create/Update RPCs. But, right now, several things rely on the ability to upsert (e.g `tctl create -f` / the terraform provider / the kubernetes operator). Once we reach a point where all of those no longer rely on this behaviour, then I think we can almost globally deprecate Upsert RPCs.

> One loose end I have on this topic is: how do Upserts behave with the new optimistic locking behavior?

IIRC the behaviour of Upsert RPCs has not been changed. Optimistic locking is entirely ignored.

I believe that in addition to this, Update RPCs will be encouraged to support update masks where appropriate. At least in this case, my UpdateBot RPC also implements an `update_mask` field which allows selective modification of the roles or traits associated with the Bot.

Reply:
Upsert is by definition unconditional, and it's required (at least as a primitive operation, it doesn't have to be exposed through grpc necessarily) for the cache to work.

The old `Update` primitive backend operation has no right to exist, and it has the semantics that @ibeckermayer mentioned. `ConditionalUpdate` is the one that you want to use, and you use exactly like that. You get the resource, you make some decisions and calculate the new resource, you `ConditionalUpdate` the resource with a constraint on the revision you got, and your update succeeds if and only if nothing has written the underlying backend item.

Whether the revision should be exposed to the client over grpc is a choice that should be done depending on the resource type and the semantics we want to give to each operation; it might make sense sometimes to expose an "update" function that lets the caller update some fields with no regard to what was there in the first place, but even in that case the auth server should 100% use the conditional update operation internally.

Reply:
(and yeah, "i just read the item successfully so the item exists" is 100% bogus and it's a pattern that we should start removing from Teleport whenever possible)

Reply:
#### @strideynet 
> It was my understanding that this was the entire motivation behind adding optimistic locking to the update RPC.

I was referring to the state of things pre-optimistic-locking.

> I believe if anything long term goal is to do the opposite of what you suggest, and eliminate upsert in favour of create and update. Right now, several things rely on the ability to upsert (e.g tctl create -f / the terraform provider / the kubernetes operator). Once we reach a point where all of those no longer rely on this behaviour, then I think we can almost globally deprecate upsert RPCs.

Seems like an odd goal... Is `create -f` not the user explicitly asking for an upsert? I can imagine implementing it with a get + update, but I'm not currently seeing what the point of that would be.

#### @espadolini 

> Upsert is by definition unconditional

Granted, dumb question.

> The old Update primitive backend operation has no right to exist, and it has the semantics that @ibeckermayer mentioned. ConditionalUpdate is the one that you want to use, and you use exactly like that. You get the resource, you make some decisions and calculate the new resource, you ConditionalUpdate the resource with a constraint on the revision you got, and your update succeeds if and only if nothing has written the underlying backend item since you read it.

Is this not essentially what `Update` has become, presuming it now uses the optimistic locking mechanism? 

I can see several plausible update-like semantics:

1. update the entire object irrespective of revision (error if object dne)
2. update the entire object on a specific revision (error if revision/object dne)
3. patch a few fields irrespective of revision (error if object dne)
4. patch a few fields on a specific revision (error if revision/object dne)
5. write an entire object irrespective of revision/existence (upsert)

1. is the status quo 
2. is presumably the primary reason we implemented optimistic locking 
3. we have in a few places (HTTP patch semantic)
4. seems practically useless -- if we have the object's revision, that typically means we have the object, so we may as well just use 2
5. seems useful to me, although there seems to be some discussion of getting rid of it

Reply:
Thanks for pointing out that PR @strideynet, this whole discussion is more suited to that RFD so let's continue it over there: https://github.com/gravitational/teleport/pull/34103#discussion_r1428615401

Reply:
@ibeckermayer do you think this blocks this PR moving forward ? I'm a little concerned about how long discussing this will run and I have some time pressure to get this merged as it's blocking other work.

Reply:
Not necessarily, I can review the rest and we can just consider this pending. 

Since you did go through the trouble of adding a field mask, it's worth following [this line of inquiry](https://github.com/gravitational/teleport/pull/34103/files/b5916b1a93a818a1b7d0667beaaf0c905ae19655#r1430639369) through to its conclusion.